velox/functions/sparksql/specialforms/GetStructField.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/specialforms/GetStructField.h"
+#include "velox/expression/ConstantExpr.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+// Returns the value of nested field in the input ``args``.
+// The input ``args`` must be of row type and nested complex type is allowed.
+// The field position is specified by ``ordinal``.
+// If ``ordinal`` is negative or greater than the children size of input,
+// it will throw exception.
+class GetStructFieldFunction : public exec::VectorFunction {
+ public:
+  GetStructFieldFunction(int32_t ordinal) : ordinal_(ordinal) {}
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    exec::LocalDecodedVector decoded(context, *args[0], rows);
+    auto rowData = decoded->base()->as<RowVector>();
+    VELOX_USER_CHECK_GE(
+        ordinal_, 0, "Invalid ordinal. Should be greater than 0.");
+    VELOX_USER_CHECK_LT(
```

Comment:
Perhaps move the checks to 'constructSpecialForm'. Also wondering if Spark also throws exception for the two cases.

Reply:
This is an internal expression and is not supposed to be used by end user directly. It's generated by `ExtractValue` and the `ordinal` is calculated by Spark internally. So there is no check in spark and should be no invalid `ordinal` value.

Reply:
In `constructSpecialForm`, we can't get the childSize of `args[0]`( which is a row vector), right? since it's not decoded. Correct me if I'm wrong

Reply:
I see. Perhaps only move the check at L40 to `constructSpecialForm`.